rust/lance/src/dataset.rs

```diff
@@ -1436,7 +1438,216 @@ impl Dataset {
     }
 }
 
+#[derive(Debug, Clone, PartialEq)]
+pub struct BatchInfo {
+    pub fragment_id: u32,
+    pub batch_index: usize,
+}
+
+pub trait UDFCache: Send + Sync {
+    fn get_batch(&self, info: &BatchInfo) -> Result<Option<RecordBatch>>;
+    fn insert_batch(&self, info: BatchInfo, batch: RecordBatch) -> Result<()>;
+    fn get_fragment(&self, fragment_id: u32) -> Result<Option<Fragment>>;
+    fn insert_fragment(&self, fragment: Fragment) -> Result<()>;
+}
+
+pub struct BatchUDF {
+    #[allow(clippy::type_complexity)]
+    pub mapper: Box<dyn Fn(&RecordBatch) -> Result<RecordBatch> + Send + Sync>,
+    /// The schema of the returned RecordBatch
+    pub append_schema: Arc<ArrowSchema>,
+    /// The names of the columns that are required to be present in the input
+    /// RecordBatch. If None, then all columns will be read and passed in.
+    pub read_columns: Option<Vec<String>>,
```

Comment:
This is similar to what pyspark PandasUDF does
https://spark.apache.org/docs/3.1.3/api/python/reference/api/pyspark.sql.functions.pandas_udf.html

In PandasUDF, it makes the caller to determine input columns, not the developer who wrote the udf.  In such way, it decouples the one who uses udf and the one who writes udf (who might be in different teams in large organizations).

Also the udf can be used for different dataset when they use alias to adjust the columns names.

For example,
we can do `add_column("image_20240101 as image", chop_image)` and `add_column("image_lowres", chop_image)`. without two implementations.

Reply:
That makes sense. I can move `read_columns` to be a parameter of `add_columns()`.